### PR TITLE
#2559 - Read Name and Reaction Conditions from RDF Metadata as a text

### PR DESCRIPTION
**Pull request by even1024**
## Generic request
- [x] PR name follows the pattern `#1234 – issue name`
- [x] branch name does not contain '#'
- [x] base branch (master or release/xx) is correct
- [x] PR is linked with the issue
- [x] task status changed to "Code review"
- [x] code follows product standards
- [x] regression tests updated

### CODE REVIEW

File: core/indigo-core/layout/src/pathway_layout.cpp [hunk 1]
```diff
@@ -142,7 +144,62 @@ void PathwayLayout::buildLayoutTree()
...
+        if (totalLines && currentLayoutItem.children.size() > 1)
+        {
+            if (totalLines > MIN_LINES_COUNT)
+                totalLines = MIN_LINES_COUNT;
+
+            auto totalHeight = std::accumulate(currentLayoutItem.children.begin(), currentLayoutItem.children.end(), 0.0f,
+                                               [](float summ, const PathwayLayoutItem* item) { return summ + item->height; });
+            totalHeight -= currentLayoutItem.children.front()->height / 2;
+            totalHeight -= currentLayoutItem.children.back()->height / 2;
+            totalHeight /= 2.0f;
```
Comment (AlexanderSavelyev) [inline core/indigo-core/layout/src/pathway_layout.cpp:158]: add constant for 2

File: core/indigo-core/layout/src/pathway_layout.cpp [hunk 2]
```diff
@@ -142,7 +144,62 @@ void PathwayLayout::buildLayoutTree()
...
+            float targetHeight = totalLines * _text_height + _reaction_margin_size;
+            if (totalHeight < targetHeight)
+            {
+                auto adjustHeight = (targetHeight - totalHeight) / (currentLayoutItem.children.size() - 1);
+                for (auto& child : currentLayoutItem.children)
+                    child->height += adjustHeight * 1.5f; // should be 2.0f
```
Comment (AlexanderSavelyev) [inline core/indigo-core/layout/src/pathway_layout.cpp:164]: add constant for 1.5

File: core/indigo-core/layout/pathway_layout.h [hunk 3]
```diff
@@ -26,10 +26,13 @@
 #include <iostream>
 #include <list>
 #include <numeric>
+#include <rapidjson/stringbuffer.h>
```
Comment (AlexanderSavelyev) [inline core/indigo-core/layout/pathway_layout.h:29]: there should be no json in layout

File: core/indigo-core/layout/pathway_layout.h [hunk 4]
```diff
@@ -239,6 +248,10 @@ namespace indigo
         void secondWalk(PathwayLayoutItem* node, PathwayLayoutItem* parent, float modifier, int depth);
 
         void applyLayout();
+        void generateKETTextBlocks(rapidjson::Writer<rapidjson::StringBuffer>& writer, const ObjArray<Array<char>>& props, const std::string& style,
```
Comment (AlexanderSavelyev) [inline core/indigo-core/layout/pathway_layout.h:251]: move the KET from layout

File: core/indigo-core/layout/pathway_layout.h [hunk 5]
```diff
@@ -239,6 +248,10 @@ namespace indigo
...
+                                   float& max_height);
+        void addKETText(PathwayReaction::ReactionNode& node, const Vec2f text_pos_bl, float text_height_limit);
```
Comment (AlexanderSavelyev) [inline core/indigo-core/layout/pathway_layout.h:253]: move the KET out of layout

File: core/indigo-core/layout/src/pathway_layout.cpp [hunk 6]
```diff
@@ -142,7 +144,62 @@ void PathwayLayout::buildLayoutTree()
...
+void PathwayLayout::copyTextPropertiesToNode(const PathwayReaction::SimpleReaction& reaction, PathwayReaction::ReactionNode& node)
+{
+    auto& props = reaction.properties;
+    // split text labels and put them into the reaction node name_text and conditions_text
+    auto text_max_width = _default_arrow_size - _reaction_margin_size * 2;
+    auto boldWidthLambda = [text_max_width](char ch) { return text_max_width / MAX_SYMBOLS; };   // TODO: implement bold font width
+    auto italicWidthLambda = [text_max_width](char ch) { return text_max_width / MAX_SYMBOLS; }; // TODO: implement italic font width
+    for (auto prop_idx = props.begin(); prop_idx != props.end(); prop_idx = props.next(prop_idx))
+    {
+        if (props.key(prop_idx) == std::string("Name"))
```
Comment (AlexanderSavelyev) [inline core/indigo-core/layout/src/pathway_layout.cpp:179]: add constant for Name

File: core/indigo-core/layout/src/pathway_layout.cpp [hunk 7]
```diff
@@ -250,28 +307,159 @@ void PathwayLayout::applyLayout()
...
             }
         }
     }
 
-    for (auto& arrow : metaArrows)
+    for (auto& arrow : metaObjects)
         _reaction.meta().addMetaObject(arrow.second.release());
 }
 
+void PathwayLayout::generateKETTextBlocks(Writer<StringBuffer>& writer, const ObjArray<Array<char>>& props, const std::string& style, float& height)
```
Comment (AlexanderSavelyev) [inline core/indigo-core/layout/src/pathway_layout.cpp:341]: move KET out of layout

File: core/indigo-core/layout/src/pathway_layout.cpp [hunk 8]
```diff
@@ -142,7 +144,62 @@ void PathwayLayout::buildLayoutTree()
...
+        {
+            auto splitted_name = splitText(props.value(prop_idx).ptr(), text_max_width, boldWidthLambda);
+            for (auto& line : splitted_name)
+            {
+                node.name_text.push().readString(line.c_str(), true);
+                node.text_width = std::max(node.text_width, std::accumulate(line.begin(), line.end(), 0.0f,
+                                                                            [&boldWidthLambda](float sum, char ch) { return sum + boldWidthLambda(ch); }));
+            }
+        }
+        else if (props.key(prop_idx) == std::string("Reaction Conditions"))
```
Comment (AlexanderSavelyev) [inline core/indigo-core/layout/src/pathway_layout.cpp:189]: add constant for Reaction conditions

**Review by AlexanderSavelyev — CHANGES_REQUESTED** [review]
add tests with rdf files

File: core/indigo-core/layout/src/pathway_layout.cpp (hunk 1 above)
Comment (even1024) [reply core/indigo-core/layout/src/pathway_layout.cpp:158]: 2 is not a magic number. There is no need to create a constant for 2.

File: core/indigo-core/layout/pathway_layout.h (hunk 3 above)
Comment (even1024) [reply core/indigo-core/layout/pathway_layout.h]: fixed

File: core/indigo-core/layout/pathway_layout.h (hunk 4 above)
Comment (even1024) [reply core/indigo-core/layout/pathway_layout.h]: fixed

File: core/indigo-core/layout/src/pathway_layout.cpp (hunk 7 above)
Comment (even1024) [reply core/indigo-core/layout/src/pathway_layout.cpp]: fixed

File: core/indigo-core/layout/src/pathway_layout.cpp (hunk 8 above)
Comment (even1024) [reply core/indigo-core/layout/src/pathway_layout.cpp:189]: fixed

File: core/indigo-core/layout/src/pathway_layout.cpp (hunk 6 above)
Comment (even1024) [reply core/indigo-core/layout/src/pathway_layout.cpp:179]: fixed

File: core/indigo-core/layout/pathway_layout.h (hunk 5 above)
Comment (even1024) [reply core/indigo-core/layout/pathway_layout.h]: fixed

File: core/indigo-core/layout/src/pathway_layout.cpp [hunk 9]
```diff
@@ -19,9 +19,9 @@
 #include <queue>
 
 #include "layout/pathway_layout.h"
-#include "molecule/ket_commons.h"
 
 using namespace indigo;
+using namespace rapidjson;
```
Comment (AlexanderSavelyev) [inline core/indigo-core/layout/src/pathway_layout.cpp:24]: is rapidjson used?

File: api/tests/integration/tests/formats/reactions/rdf/pathway12.rdf [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+   59.9745  -68.0415    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
+   59.1089  -67.5406    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
+   58.2425  -68.0400    0.0000 Cl  0  0  0  0  0  0  0  0  0  0  0  0
+  1  2  2  0  0  0  0
+  2  3  1  0  0  0  0
+  3  4  1  0  0  0  0
+M  END
+
+$DTYPE Name
+$DATUM ABCDEFGHIJKLMNOPQRSTUVWXYZABCDABCDEFGHIJKLMNOPQRSTUVWXYZABCDABCDEFGHIJKLMNOPQRSTUVWXYZABCDABCDEFGHIJKLMNOPQRSTUVWXYZABCDABCDEFGHIJKLMNOPQRSTUVWXYZABCDABCDEFGHIJKLMNOPQRSTUVWXYZABCDABCDEFGHIJKLMNOPQRSTUVWXYZABCD
```
Comment (AlexanderSavelyev) [inline api/tests/integration/tests/formats/reactions/rdf/pathway12.rdf:450]: no test case for "Not available"

**Review by AlexanderSavelyev — CHANGES_REQUESTED** [review]

**Review by AlexanderSavelyev — APPROVED** [review]

**Pull request merged**